### PR TITLE
Remove Parent from IR nodes

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/InstrumentationPass.cs [hunk 1]
```diff
@@ -45,33 +42,33 @@ private static void AddInstrumentation(InstrumentationItem item)
...
-            item.Node.Parent.Children.Insert(nodeIndex + 2, endNode);
+            var nodeIndex = item.Parent.Children.IndexOf(item.Node);
+            item.Parent.Children.Insert(nodeIndex, beginNode);
+            item.Parent.Children.Insert(nodeIndex + 2, endNode);
         }
 
         private struct InstrumentationItem
         {
-            public InstrumentationItem(RazorIRNode node, bool isLiteral, SourceSpan source)
+            public InstrumentationItem(RazorIRNode node, RazorIRNode parent, bool isLiteral, SourceSpan source)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/InstrumentationPass.cs]: This is the new basic pattern for rewriters - grab the parent and the node during the visiting phase, and then act on it later.

File: src/Microsoft.AspNetCore.Razor.Language/DefaultDirectiveIRPass.cs [hunk 2]
```diff
@@ -20,20 +20,19 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIRNo
             walker.VisitDocument(irDocument);
 
             var classNode = walker.ClassNode;
-            foreach (var node in walker.FunctionsDirectiveNodes)
+            foreach (var (node, parent) in walker.FunctionsDirectiveNodes)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultDirectiveIRPass.cs]: This class shows that same basic pattern but with valuetuples

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultDirectiveIRPass.cs]: I've actually got a super cool idea about this, I'll send another PR

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultDirectiveIRPass.cs]: Generic types instead of tuples?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/DefaultDirectiveIRPass.cs]: See my other PR

**Comment by rynowak** [comment]
> Only concern I have is of the repeated value tuple patterns everywhere. Leads me to think there's a better alternative.

See my next PR in this series

**Comment by rynowak** [comment]
I'm going to go ahead and merge this, but leave https://github.com/aspnet/Razor/pull/1420 in case we decide to address your crazy feedback from that one.

**Pull request closed without merging**